apps/site/pages/en/learn/asynchronous-work/discover-promises-in-nodejs.md

```diff
@@ -0,0 +1,401 @@
+
+In addition to Promises, Node.js provides several other mechanisms for scheduling tasks in the event loop.
+
+### **`queueMicrotask()`**
+
+`queueMicrotask()` is used to schedule a microtask, which is a lightweight task that runs after the currently executing script but before any other I/O events or timers. Microtasks include tasks like Promise resolutions and other asynchronous operations that are prioritized over regular tasks.
+
+```js
+queueMicrotask(() => {
+  console.log('Microtask is executed');
+});
+
+console.log('Synchronous task is executed');
+```
+
+In the above example, "Microtask is executed" will be logged after "Synchronous task is executed," but before any I/O operations like timers.
+
+### **`process.nextTick()`**
+
+`process.nextTick()` is used to schedule a callback to be executed immediately after the current operation completes. This is useful for situations where you want to ensure that a callback is executed as soon as possible, but still after the current execution context.
+
+```js
+process.nextTick(() => {
+  console.log('Next tick callback');
+});
+
+console.log('Synchronous task executed');
+```
+
+### **`setImmediate()`**:
+
+`setImmediate()` is used to execute a callback after the current event loop cycle finishes and all I/O events have been processed. This means that `setImmediate()` callbacks run after any I/O callbacks, but before timers.
+
+```js
+setImmediate(() => {
+  console.log('Immediate callback');
+});
+
+console.log('Synchronous task executed');
+```
+
+### **When to Use Each**:
+
+- Use `queueMicrotask()` for tasks that need to run immediately after the current script and before any I/O or timer callbacks, typically for Promise resolutions.
+- Use `process.nextTick()` for tasks that should execute before any I/O events, often useful for deferring operations or handling errors synchronously.
+- Use `setImmediate()` for tasks that should run after I/O events but before timers.
+
+Because these tasks execute outside of the current synchronous flow, uncaught exceptions inside these callbacks won't be caught by surrounding `try/catch` blocks and may crash the application if not properly managed (e.g., by attaching `.catch()` to Promises or using global error handlers like `process.on('uncaughtException')`).
+
+In short, the execution order is as follows:
+
+1. **Synchronous code** (e.g., regular function calls and script code)
+2. **`process.nextTick()`** callbacks
+3. **Microtasks** (e.g., Promises, `queueMicrotask()`)
+   If a microtask queues another microtask, the new one is also executed before moving on.
+4. **Timers** (e.g., `setTimeout()`, `setInterval()`)
+5. **I/O callbacks** (e.g., network and file system operations)
+   Some operations, like `close` and `error` events, may use `process.nextTick()` internally and are executed in the next tick.
+6. **`setImmediate()`** callbacks
+
+`process.nextTick()` can execute at any stage in the execution order, as it is not bound by the event loop.
```

Comment:
Sorry if I wasn't clear... this is *also* true of draining the microtask queue. We always drain both the next tick and microtask queue (in that order) at the same time.

Reply:
I realized we have another article on the event loop, so I've referenced it instead